tfinstall/tfinstall.go

```diff
@@ -0,0 +1,346 @@
+	}
+	defer os.RemoveAll(sumsTmpDir)
+
+	sumsFilename := "terraform_" + tfVersion + "_SHA256SUMS"
+	sumsSigFilename := sumsFilename + ".sig"
+
+	sumsUrl := fmt.Sprintf("%s/%s/%s",
+		releaseHost, tfVersion, sumsFilename)
+	sumsSigUrl := fmt.Sprintf("%s/%s/%s",
+		releaseHost, tfVersion, sumsSigFilename)
+
+	client.Src = sumsUrl
+	client.Dst = sumsTmpDir
+	err = client.Get()
+	if err != nil {
+		return "", fmt.Errorf("error fetching checksums: %s", err)
+	}
+
+	client.Src = sumsSigUrl
+	err = client.Get()
+	if err != nil {
+		return "", fmt.Errorf("error fetching checksums signature: %s", err)
+	}
+
+	sumsPath := filepath.Join(sumsTmpDir, sumsFilename)
+	sumsSigPath := filepath.Join(sumsTmpDir, sumsSigFilename)
+
+	err = verifySumsSignature(sumsPath, sumsSigPath)
+	if err != nil {
+		return "", err
+	}
+
+	// secondly, download Terraform itself, verifying the checksum
+	url := tfUrl(tfVersion, osName, archName)
+	client.Src = url
+	client.Dst = tfDir
+	client.Mode = getter.ClientModeDir
+	err = client.Get()
+	if err != nil {
+		return "", err
+	}
+
+	return filepath.Join(tfDir, "terraform"), nil
+}
+
+func tfUrl(tfVersion, osName, archName string) string {
+	sumsFilename := "terraform_" + tfVersion + "_SHA256SUMS"
+	sumsUrl := fmt.Sprintf("%s/%s/%s",
+		releaseHost, tfVersion, sumsFilename)
+	return fmt.Sprintf(
+		"%s/%s/terraform_%s_%s_%s.zip?checksum=file:%s",
+		releaseHost, tfVersion, tfVersion, osName, archName, sumsUrl,
+	)
+}
+
+func latestVersion() (string, error) {
+	resp, err := checkpoint.Check(&checkpoint.CheckParams{
+		Product: "terraform",
+	})
+	if err != nil {
+		return "", err
+	}
+
+	if resp.CurrentVersion == "" {
+		return "", fmt.Errorf("could not determine latest version of terraform using checkpoint: CHECKPOINT_DISABLE may be set")
+	}
+
+	return resp.CurrentVersion, nil
+}
+
+// verifySumsSignature downloads SHA256SUMS and SHA256SUMS.sig and verifies
+// the signature using the HashiCorp public key.
+func verifySumsSignature(sumsPath, sumsSigPath string) error {
+	el, err := openpgp.ReadArmoredKeyRing(strings.NewReader(hashicorpPublicKey))
+	if err != nil {
+		return err
+	}
+	data, err := os.Open(sumsPath)
+	if err != nil {
+		return err
+	}
+	sig, err := os.Open(sumsSigPath)
+	if err != nil {
+		return err
+	}
+	_, err = openpgp.CheckDetachedSignature(el, data, sig)
+
+	return err
+}
+
+// verifySums verifies the hash of the downloaded file against the appropriate line
+// in SHA256SUMS
+func verifySums(terraformPath, sumsPath string) error {
+	t, err := os.Open(terraformPath)
+	if err != nil {
+		return err
+	}
+	defer t.Close()
+
+	h := sha256.New()
+	if _, err := io.Copy(h, t); err != nil {
+		return err
+	}
+
+	f, err := os.Open(sumsPath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	rd := bufio.NewReader(f)
+	for {
+		line, err := rd.ReadString('\n')
+		if err != nil {
+			if err != io.EOF {
+				return err
+			}
+			break
+		}
+		parts := strings.Fields(line)
+		if parts[0] == "terraform" {
```

Comment:
I may be misunderstanding the intentions here, but isn't the first part of checksum file the hash? e.g.

```
893050bcfc5e7445acd3a30f1500227b989b29cbd958ca64a8233589194a198d  terraform_0.12.28_darwin_amd64.zip
7f443c80d4793d3181cc579ef4c37a7cf76b63b316670b1abd7598ebe421dd44  terraform_0.12.28_freebsd_386.zip
9d3381adc015cda84687ccc4aed7714645857f901c301d7f5d20ecc78844f3fd  terraform_0.12.28_freebsd_amd64.zip
c8c76a215c7f088a547fdb318ff08a48ad45a6995a4b5b36cfcf05d01ed7c86d  terraform_0.12.28_freebsd_arm.zip
1c489282d86b16f2d5f89f38071c6dabd948a4ca7cc4e42915e604b82564f3a6  terraform_0.12.28_linux_386.zip
be99da1439a60942b8d23f63eba1ea05ff42160744116e84f46fc24f1a8011b6  terraform_0.12.28_linux_amd64.zip
ee1835429b681460e0adcc6f873a1d1390264647bae77151ba8211610dc86d83  terraform_0.12.28_linux_arm.zip
c5c32dc1896188a17f93e0e7aa3f160fe842ad0eb4e4761345e728b1ac26e8e2  terraform_0.12.28_openbsd_386.zip
d1818c9d5414b8fa5ed9618bf00abd605b016be7812e5b71661337693a47d165  terraform_0.12.28_openbsd_amd64.zip
a904775f596895ff84dcc76ffff68bebf0c115b39e21f9a612431fcadbc814da  terraform_0.12.28_solaris_amd64.zip
a6c8c2f4d20c38f3467cbc65e3d34d446751a675f4b1bfdf6be155baf14f8963  terraform_0.12.28_windows_386.zip
cd524b5b6b7cd9eec9c4e49aa37cbcb34ed1395876c212f6dde84c6e57d6ce1c  terraform_0.12.28_windows_amd64.zip

```